packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/mod.rs

```diff
@@ -276,36 +276,29 @@ impl<C: AsClient> EntityStore for PostgresStore<C> {
                 .change_context(UpdateError)?,
         );
 
-        // TODO - address potential race condition
-        //  https://app.asana.com/0/1202805690238892/1203201674100967/f
-
-        let previous_entity: PersistedEntity = transaction
-            .read_one(&Filter::for_latest_entity_by_entity_id(entity_id))
+        transaction
+            .lock_entity_for_update(entity_id)
             .await
             .change_context(UpdateError)?;
 
-        if !transaction
-            .contains_entity(entity_id)
-            .await
-            .change_context(UpdateError)?
-        {
-            return Err(Report::new(EntityDoesNotExist)
-                .attach_printable(entity_id)
-                .change_context(UpdateError));
-        }
+        // TODO - address potential race condition.
+        //  Note that the race condition may have been sorted out with the links rewrite.
+        //  https://app.asana.com/0/1202805690238892/1203201674100967/f
```

Comment:
I did, and it did work as expected.
But I would still like for us to make this check explicitly and separately from this task. The test case i reran was similar to the original minimal repro. steps, so it should have been addressed.

I am leaving the todo so we can think about this when we merge, potentially making sure that the race condition is gone after more code has been rewritten for links.

Reply:
Gotcha, I thought it might have been to have the discussion and explicit check. 

I guess I'm saying, do we want to do that as part of merging this PR and then remove the TODO, otherwise this will just sit in the backlog and we'll have to context switch to get it again.

Reply:
> I am leaving the todo so we can think about this when we merge

Oops I didn't read properly 😁, glad we're on the same page

Reply:
To update on this, I tested this again in-API rather than in-DB.
The testing methodology was to force a concurrent but ordered transaction by making one of the transactions artificially take longer. The findings are as follows:

Say we are archiving + updating entity `A` at the same time.
One (slow) request archives the entity and right after, a request to update the entity is carried out (blocked by the slow request):
- The first writer wins, which is expected with locking. We can't really do much about this, but fortunately it behaves as expected
- This means the archival is carried out
- The update fails as it would if the entity did not exist (it is now archived, so it cannot be updated.)

Flipping the operation order (and making the update wait instead of archival)
- The archival DOES NOT go through when waiting for the lock release of the `update_entity`! This means the query will act as if the latest version of the entity does not exist
- This isn't a data race problem as subsequent queries would suceed, but in the case of multiple queries locking on the same entity_id, we'd have a problem with the archival not going through in this very specific case.
- We _could maybe_ get around this issue by using `Serializable` isolation level. But we probably do not want to experiement with this right now. I think we should keep this todo

I also got rid of the nested transaction in archive here which I didn't notice before. 122dacc20316586ad896a4a7a60b5767b0de125f

cc: @TimDiekmann

Reply:
> I also got rid of the nested transaction in archive here which I didn't notice before. https://github.com/hashintel/hash/commit/122dacc20316586ad896a4a7a60b5767b0de125f

Well that's relevant to recent discussions haha, oops!

Reply:
> I think we should keep this todo

Makes sense. Can you capture these new findings in the asana task too? And maybe a link to a gist or a commit or something on how you tested this

Reply:
All info in https://app.asana.com/0/1201095311341924/1203201674100967/f